File: scripts/compiler/container.ts
/**
 * Dependency Injection Container for the Compiler Engine
 */

import * as fs from "node:fs";
import type { Highlighter } from "shiki";
import { Logger, LogLevel } from "../core/logger.ts";
import { CompilationContext } from "./Context.ts";
import { MarkdownRenderer } from "./Renderer.ts";
import type { CompilerConfig } from "./types.ts";

// ─── Service Interfaces ───────────────────────────────────────────────────

export interface IFileSystem {
  read(path: string): string;
  write(path: string, content: string): void;
  exists(path: string): boolean;
  mkdir(path: string, options?: { recursive: boolean }): void;
  rm(path: string, options?: { recursive: boolean; force: boolean }): void;
}

export interface CompilerContainer {
  config: CompilerConfig;
  logger: Logger;
  context: CompilationContext;
  renderer: MarkdownRenderer;
  fs: IFileSystem;
}

// ─── Default Implementations ──────────────────────────────────────────────

export const createDefaultFileSystem = (): IFileSystem => ({
  read: (p) => fs.readFileSync(p, "utf-8"),
  write: (p, c) => fs.writeFileSync(p, c, "utf-8"),
  exists: (p) => fs.existsSync(p),
  mkdir: (p, o) => fs.mkdirSync(p, o),
  rm: (p, o) => fs.rmSync(p, o),
});

// ─── Container Builder ────────────────────────────────────────────────────

export interface ContainerOptions {
  config: CompilerConfig;
  highlighter?: Highlighter;
  logger?: Logger;
  fs?: IFileSystem;
}

/**
 * Build a compiler service container
 */
export function createCompilerContainer(options: ContainerOptions): CompilerContainer {
  const logger = options.logger ?? new Logger(LogLevel.Info);
  const fileSystem = options.fs ?? createDefaultFileSystem();
  const context = new CompilationContext(options.config);
  const renderer = new MarkdownRenderer(options.highlighter);

  return {
    config: options.config,
    logger,
    context,
    renderer,
    fs: fileSystem,
  };
}

File: scripts-rs/src/compiler/container.rs
use crate::compiler::context::{CompilationContext, CompilerConfig};
use crate::core::Logger;
use std::fs;
use std::path::Path;

pub trait IFileSystem {
    fn read(&self, path: &Path) -> anyhow::Result<String>;
    fn write(&self, path: &Path, content: &str) -> anyhow::Result<()>;
    fn exists(&self, path: &Path) -> bool;
    fn mkdir(&self, path: &Path, recursive: bool) -> anyhow::Result<()>;
    fn rm(&self, path: &Path, recursive: bool, force: bool) -> anyhow::Result<()>;
}

pub struct DefaultFileSystem;
impl IFileSystem for DefaultFileSystem {
    fn read(&self, path: &Path) -> anyhow::Result<String> {
        fs::read_to_string(path).map_err(Into::into)
    }
    fn write(&self, path: &Path, content: &str) -> anyhow::Result<()> {
        fs::write(path, content).map_err(Into::into)
    }
    fn exists(&self, path: &Path) -> bool {
        path.exists()
    }
    fn mkdir(&self, path: &Path, recursive: bool) -> anyhow::Result<()> {
        if recursive {
            fs::create_dir_all(path).map_err(Into::into)
        } else {
            fs::create_dir(path).map_err(Into::into)
        }
    }
    fn rm(&self, path: &Path, recursive: bool, _force: bool) -> anyhow::Result<()> {
        if recursive {
            fs::remove_dir_all(path).map_err(Into::into)
        } else if path.is_dir() {
            fs::remove_dir(path).map_err(Into::into)
        } else {
            fs::remove_file(path).map_err(Into::into)
        }
    }
}

pub struct CompilerContainer {
    pub config: CompilerConfig,
    pub logger: Logger,
    pub context: CompilationContext,
    pub fs: Box<dyn IFileSystem>,
}

impl CompilerContainer {
    pub fn new(config: CompilerConfig, fs: Option<Box<dyn IFileSystem>>) -> Self {
        let context = CompilationContext::from_config(config.clone());
        let logger = Logger::new();
        let fs = fs.unwrap_or_else(|| Box::new(DefaultFileSystem));
        
        Self {
            config,
            logger,
            context,
            fs,
        }
    }
}
